### ISSUE TITLE
URL strategy not working trying to remove hashtag(#) from PWA

### ISSUE THREAD

**Issue by KabluBR** [flutter/flutter#69254]
## Steps to Reproduce

To reproduce the bug, I just tried to adapt an existing app to web PWA and removing the # hashtag using this:

[STEPS TO REMOVE # FROM URL](https://github.com/flutter/flutter/issues/33245#issuecomment-705095853)

**Expected results:** url without the /#/ in my pwa

**Actual results:** crash with the following results:

## The error/bug:
```
TypeError: dart.global._flutter_web_set_location_strategy is not a function
at Object.setUrlStrategy (http://localhost:29624/packages/flutter_web_plugins/src/navigation/js_url_strategy.dart.lib.js:243:17)
at Object.configureApp (http://localhost:29624/packages/project/configure_web.dart.lib.js:10:18)
```

FLUTTER DOCTOR -V

```
flutter doctor -v
[√] Flutter (Channel master, 1.24.0-4.0.pre.123, on Microsoft Windows [Version 10.0.18363.1139], locale pt-BR)
    • Flutter version 1.24.0-4.0.pre.123 at c:\DEV\flutter
    • Framework revision 735f6eec2c (8 minutes ago), 2020-10-28 15:09:46 -0700
    • Engine revision 64e6599910
    • Dart version 2.11.0 (build 2.11.0-260.0.dev)

[√] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at C:\android-sdk-win
    • Platform android-29, build-tools 29.0.2
    • ANDROID_HOME = C:\android-sdk-win
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Android Studio (version 4.1.0)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] VS Code, 64-bit edition (version 1.49.2)
    • VS Code at C:\Program Files\Microsoft VS Code
    • Flutter extension version 3.15.0

[√] Connected device (3 available)
    • Web Server (web) • web-server • web-javascript • Flutter Tools
    • Chrome (web)     • chrome     • web-javascript • Google Chrome 86.0.4240.111
    • Edge (web)       • edge       • web-javascript • Microsoft Edge 86.0.622.56

• No issues found!
```

**Comment by Sun3** [comment]
I am having the same error:

```
[√] Flutter (Channel master, 1.24.0-4.0.pre.122, on Microsoft Windows [Version 10.0.19041.572], locale en-US)
    • Flutter version 1.24.0-4.0.pre.122 at C:\Users\marco\flutter
    • Framework revision 4aad058a49 (81 minutes ago), 2020-10-28 17:02:37 -0400
    • Engine revision 64e6599910
    • Dart version 2.11.0 (build 2.11.0-260.0.dev)

[√] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at C:\Users\marco\AppData\Local\Android\sdk
    • Platform android-30, build-tools 30.0.2
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.7.6)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.7.30611.23
    • Windows 10 SDK version 10.0.19041.0

[√] Android Studio (version 4.1.0)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] VS Code (version 1.50.1)
    • VS Code at C:\Users\marco\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.15.1

[√] Connected device (4 available)
    • Windows (desktop) • windows    • windows-x64    • Microsoft Windows [Version 10.0.19041.572]
    • Web Server (web)  • web-server • web-javascript • Flutter Tools
    • Chrome (web)      • chrome     • web-javascript • Google Chrome 86.0.4240.111
    • Edge (web)        • edge       • web-javascript • Microsoft Edge 86.0.622.51

• No issues found!
```

**Comment by TahaTesser** [comment]
This doesn't reproduce in a new project, given the reproduce case is for existing projects, makes sense 
https://eager-agnesi-6e5382.netlify.app/
<details> 
<summary>flutter doctor -v</summary> 

```bash
[✓] Flutter (Channel master, 1.24.0-4.0.pre.138, on Mac OS X 10.15.7 19H2
    darwin-x64, locale en-GB)
    • Flutter version 1.24.0-4.0.pre.138 at /Users/tahatesser/Code/flutter_master
    • Framework revision 13896b3bd1 (7 hours ago), 2020-10-28 19:43:19 -0700
    • Engine revision 073263e39d
    • Dart version 2.11.0 (build 2.11.0-260.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.1, Build version 12A7403
    • CocoaPods version 1.10.0.rc.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.50.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.15.1

[✓] Connected device (4 available)
    • RMX2001 (mobile) • EUYTFEUSQSRGDA6D • android-arm64  • Android 10 (API 29)
    • macOS (desktop)  • macos            • darwin-x64     • Mac OS X 10.15.7
      19H2 darwin-x64
    • Web Server (web) • web-server       • web-javascript • Flutter Tools
    • Chrome (web)     • chrome           • web-javascript • Google Chrome
      86.0.4240.111

• No issues found!
```

</details>

**Comment by Sun3** [comment]
I just tested by creating a new Flutter Project, then added the instructions [from](https://github.com/flutter/flutter/issues/33245#issuecomment-705095853) I still receive the same error:

Perhaps are you using a different method?

```
Running "flutter pub get" in flutter_web_test...
Launching lib\main.dart on Edge in debug mode...
Syncing files to device Edge...
Debug service listening on ws://127.0.0.1:64938/zXIjQwNWpag=/ws
Debug service listening on ws://127.0.0.1:64938/zXIjQwNWpag=/ws
TypeError: dart.global._flutter_web_set_location_strategy is not a function
    at Object.setUrlStrategy (http://localhost:64860/packages/flutter_web_plugins/src/navigation/js_url_strategy.dart.lib.js:243:17)
    at Object.configureApp (http://localhost:64860/packages/flutter_web_test/configure_web.dart.lib.js:10:18)
    at main$ (http://localhost:64860/packages/flutter_web_test/main.dart.lib.js:504:19)
    at main (http://localhost:64860/web_entrypoint.dart.lib.js:31:27)
    at main.next (<anonymous>)
    at http://localhost:64860/dart_sdk.js:37902:33
    at _RootZone.runUnary (http://localhost:64860/dart_sdk.js:37756:58)
    at _FutureListener.thenAwait.handleValue (http://localhost:64860/dart_sdk.js:32717:29)
    at handleValueCallback (http://localhost:64860/dart_sdk.js:33265:49)
    at Function._propagateToListeners (http://localhost:64860/dart_sdk.js:33303:17)
    at async._AsyncCallbackEntry.new.callback (http://localhost:64860/dart_sdk.js:33028:27)
    at Object._microtaskLoop (http://localhost:64860/dart_sdk.js:38017:13)
    at _startMicrotaskLoop (http://localhost:64860/dart_sdk.js:38023:13)
    at http://localhost:64860/dart_sdk.js:33520:9
```

Thank you.

**Comment by sashker** [comment]
Having the same issue on the existing project (created very recently, though)
TypeError: dart.global._flutter_web_set_location_strategy is not a function

Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel master, 1.24.0-4.0.pre.138, on Mac OS X 10.15.7 19H2 darwin-x64, locale ru)
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
[✓] Chrome - develop for the web
[✓] Android Studio (version 4.0)
[✓] VS Code
[✓] Connected device (3 available)

**Comment by rezajalilian** [comment]
I've been testing it for a few hours in realease mode, it works properly but if you try to access straight to the url, it just fails.
If I try to access a page straight from the browser, the router can't find it(Tested in a real server).
For example if you access "testweb.com" it works but if you access "testweb.com/loginpage" server returns 404 error, it doesn't happen with hashtag.

**Comment by TahaTesser** [comment]
Hi @Sun3 
Followed steps from https://github.com/flutter/flutter/issues/33245#issuecomment-705095853 in a new project  https://github.com/flutter/flutter/issues/69254#issuecomment-718529383, had no issues o Windows

**Comment by KabluBR** [comment]
One more update on the error...

I was not posting this but it was a little strange... anyway...

Situation 1
If I click to run the debug in Android Studio, it opens the chrome and in some seconds the mentioned error appear...

Situation 2
If I click debug and when it opens the browser chrome I quickly open inspector and toggle the view to mobile view (device view) before the project loads, the error does not fire... it will just fire if I click save on Android Studio and it reloads...

**Comment by Sun3** [comment]
@TahaTesser Thanks for posting the hello sample. I ran it from my computer and received the same error as before.

I am pasting the error log from the browser this time showing that the ```url_strategy.dart``` file function ```void setUrlStrategy(UrlStrategy strategy){}``` calling ```jsSetUrlStrategy(convertToJsUrlStrategy(strategy));``` is causing the error.

url_strategy.dart file
```
/// Change the strategy to use for handling browser URL.
///
/// Setting this to null disables all integration with the browser history.
void setUrlStrategy(UrlStrategy strategy) {
  jsSetUrlStrategy(convertToJsUrlStrategy(strategy));
}
```

Browser console error log:

```
url_strategy.dart:18 Uncaught (in promise) TypeError: dart.global._flutter_web_set_location_strategy is not a function
    at Object.setUrlStrategy (url_strategy.dart:18)
    at Object.configureApp (configure_web.dart:4)
    at main$ (main.dart:5)
    at main (web_entrypoint.dart:16)
    at main.next (<anonymous>)
    at async_patch.dart:45
    at _RootZone.runUnary (zone.dart:1450)
    at _FutureListener.thenAwait.handleValue (future_impl.dart:143)
    at handleValueCallback (future_impl.dart:696)
    at Function._propagateToListeners (future_impl.dart:725)
    at async._AsyncCallbackEntry.new.callback (future_impl.dart:393)
    at Object._microtaskLoop (schedule_microtask.dart:41)
    at _startMicrotaskLoop (schedule_microtask.dart:50)
    at async_patch.dart:166
```

**Comment by matsp** [comment]
Maybe this helps. The issue just appeared in my codebase too after upgrading from **1.24-3.0.pre** to **1.24-6.0.pre** (on dev channel).

**Comment by KabluBR** [comment]
@mdebbar  It worked!! Thank you!!!

**Comment by matsp** [comment]
Now the **beta** channel with version **1.24.0-10.2.pre** is also broken:

`TypeError: dart.global._flutter_web_set_location_strategy is not a function
    at Object.setUrlStrategy (http://localhost:49988/packages/flutter_web_plugins/src/navigation/js_url_strategy.dart.lib.js:281:17)`

`
Flutter 1.24.0-10.2.pre • channel beta • https://github.com/flutter/flutter.git
Framework • revision 022b333a08 (vor 3 Tagen) • 2020-11-18 11:35:09 -0800
Engine • revision 07c1eed46b
Tools • Dart 2.12.0 (build 2.12.0-29.10.beta)
`

**Comment by agreaves** [comment]
I'm still seeing this issue in the beta channel. Is this fix available for beta?

**Comment by NaikSoftware** [comment]
Still not fixed in beta

**Comment by ponderMuse** [comment]
Getting same error. I'm on Flutter 1.24.0-10.2.pre • channel beta

**Comment by ponderMuse** [comment]
@matsp Switched to channel master and can confirm the issue has now gone away. Cheers

**Comment by DarthSid12** [comment]
I did not exactly understand what @mdebbar suggested. I am still getting the error 
`TypeError: dart.global._flutter_web_set_location_strategy is not a function`
I have followed all the steps till here. I am in master channel. Do I need to add something else?

**Comment by gibahjoe** [comment]
Please when is this getting to beta? beta was fine before I ran flutter upgrade.

Now I am getting `dart.global._flutter_web_set_location_strategy` is not a function. It works well on master tho

**Comment by matsp** [comment]
> Please when is this getting to beta? beta was fine before I ran flutter upgrade.
> 
> Now I am getting `dart.global._flutter_web_set_location_strategy` is not a function. It works well on master tho

You can also go back to the version that works for you and upgrade if this patch is coming to beta.

**Comment by matsp** [comment]
The fix is now available on the dev channel aka `1.25.0-4.0.pre` at the moment 👍
